docs: Enhance Configuration Overview for Bacalhau node (#268)
bacalhau-project/docs.bacalhau.org: docs/running-node/configuration.md

---
sidebar_label: 'Configuration'
sidebar_position: 160

> description: How to configure your Bacalhau node.

Author: I was thinking this was said document. What would we expect in a document targeted at end-users? Maybe a heavier focus on client commands?

> description: How to configure your Bacalhau node.

Reviewer: I think so. we have a CLI reference page as well and it may be worth bringing that up to date, although with the upcoming CLI changes I'm not sure.

---

# Configuration Overview

Bacalhau employs the [viper](https://github.com/spf13/viper) and [cobra](https://github.com/spf13/cobra) libraries for configuration management. Users can configure their Bacalhau node through a combination of command-line flags, environment variables, and the dedicated configuration file.

## The Bacalhau Repo

> Bacalhau manages its configuration, metadata, and internal state within a specialized repository named `.bacalhau`. Serving as the heart of the Bacalhau node, this repository holds the data and settings that determine node behavior. It's located on the user's filesystem, and by default, Bacalhau initializes this repository at the `$HOME/.bacalhau` location.

Author: d832311b2c49b56db83c070394c7deaa968c6f8e

To customize this location, users can:

1. Set the `BACALHAU_DIR` environment variable.

> 2. Utilize the `--repo` flag and specify their desired path.

Reviewer: ```suggestion
2. Utilize the `--repo` command line flag to specify their desired path.
```

Upon executing a Bacalhau command for the first time, the system will initialize the `.bacalhau` repository. If such a repository already exists, Bacalhau will seamlessly access its contents.

> **Structure of a Newly Initialized `.bacalhau`** Repository

Reviewer 2: The bold is confusing, this should be a heading.

Author: done via d832311b2c49b56db83c070394c7deaa968c6f8e

Below is the structure of a freshly initialized `.bacalhau` repository:

```shell
$ tree ~/.bacalhau
├── QmdGUjsMHEgtAfdtw7U62yPEcAZFtA33tKMsczLToegZtv-compute
│   ├── executions.db
│   └── jobStats.json
├── QmdGUjsMHEgtAfdtw7U62yPEcAZFtA33tKMsczLToegZtv-requester
│   └── jobs.db
├── config.yaml

> ├── executor_storages

Reviewer: ```suggestion
├── executor_storages/
```

├── libp2p_private_key

> ├── plugins

Reviewer: ```suggestion
├── plugins/
```

├── repo.version
└── user_id.pem
```

This repository comprises four directories and seven files:

> - - **Files**:

Reviewer: `###` ?

  
    1. `user_id.pem`:

>        - This file houses the Bacalhau node's user private key, used for signing requests to a Requester Node.
>        - Format: RSA private key.

Reviewer: ```suggestion
       - This file houses the Bacalhau node user's cryptographic private key, used for signing requests sent to a Requester Node.
       - Format: PEM
```

    2. `repo.version`:
       - Indicates the version of the Bacalhau node's repository.
       - Format: JSON, e.g., `{"Version":1}`.
    3. `libp2p_private_key`:
       - Stores the Bacalhau node's [libp2p](https://libp2p.io/) private key, essential for its network identity. The NodeID of a Bacalhau node is derived from this key.
       - Format: Base64 encoded RSA private key.
    4. `config.yaml`:
       - Contains configuration settings for the Bacalhau node.
       - Format: YAML.
  

>     **Directories**:

Reviewer: `###` ?

  
    1. `QmdGUjsMHEgtAfdtw7U62yPEcAZFtA33tKMsczLToegZtv-compute`:

>        - Houses the [BoltDB](https://github.com/etcd-io/bbolt) `executions.db` database, which aids the Compute node in state persistence. Additionally, the `jobStats.json` file records the Compute Node's completed jobs tally.

Reviewer: ```suggestion
       - Contains the [BoltDB](https://github.com/etcd-io/bbolt) `executions.db` database, which aids the Compute node in state persistence. Additionally, the `jobStats.json` file records the Compute Node's completed jobs tally.
```

       - Note: The segment `QmdGUjsMHEgtAfdtw7U62yPEcAZFtA33tKMsczLToegZtv` is a unique NodeID for each Bacalhau node, derived from the `libp2p_private_key`.
    2. `QmdGUjsMHEgtAfdtw7U62yPEcAZFtA33tKMsczLToegZtv-requester`:
       - Contains the [BoltDB](https://github.com/etcd-io/bbolt) `jobs.db` database for the Requester node's state persistence.
       - Note: NodeID derivation is similar to the Compute directory.
    3. `executor_storages`:
       - Storage for data handled by Bacalhau storage drivers.
    4. `plugins`:
       - Houses binaries that allow the Compute node to execute specific tasks.
       - Note: This feature is currently experimental and isn't active during standard node operations.

## Configuring a Bacalhau Node

Within a `.bacalhau` repository, a `config.yaml` file may be present. This file serves as the configuration source for the bacalhau node and adheres to the YAML format. 

**Note:** On the initialization of a new `.bacalhau` repository, Bacalhau will generate a `config.yaml` file. However, if Bacalhau is opening an existing repository, it will not create this file if it's absent.

>  Although the `config.yaml` file is optional, its presence allows Bacalhau to load custom configurations; otherwise, Bacalhau operates with its built-in default values. 

Reviewer: ```suggestion
 Although the `config.yaml` file is optional, its presence allows Bacalhau to load custom configurations; otherwise, Bacalhau is configured with built-in default values, environment variables and command line flags.
```

Author: oooo yeah good catch.

> Modifications to the `config.yaml` file will not be dynamically loaded by the Bacalhau node. A restart of the node is imperative for any changes to take effect. Bacalhau determines its configuration based on the following precedence order, with each item superseding the subsequent:

Reviewer: ```suggestion
Modifications to the `config.yaml` file will not be dynamically loaded by the Bacalhau node. A restart of the node is required for any changes to take effect. Bacalhau determines its configuration based on the following precedence order, with each item superseding the subsequent:
```

> 1. Flag

Reviewer: ```suggestion
1. Command-line flag
```

2. Environment Variable
3. Config File
4. Defaults

> **Relationship Between `config.yaml` and Bacalhau Environment Variables** 

Reviewer 2: Another heading

Author: d832311b2c49b56db83c070394c7deaa968c6f8e

Bacalhau establishes a direct relationship between the leaf keys within the `config.yaml` file and corresponding environment variables. For leaf keys, the environment variable name is constructed by capitalizing each segment of the key, and then joining them with underscores, prefixed with `BACALHAU_`. 

> For example, a YAML leaf key with the path `Node.Ipfs.Connect` translates to the environment variable `BACALHAU_NODE_IPFS_CONNECT`. 

Reviewer 2: This is the first time we've talked about leaf keys, and I don't know what this means.

Author: This is a bit obtuse, and I wasn't sure how best to word it. This is what I am trying to communicate, I'll use an example from our config:
```yaml
Node:
    ClientAPI:
        Host: bootstrap.production.bacalhau.org
        Port: 1234
        TLS:
            AutoCert: ""
            AutoCertCachePath: ""
            ServerCertificate: ""
            ServerTLSKey: ""
    ServerAPI:
        Host: 0.0.0.0
        Port: 1234
        TLS:
            AutoCert: ""
            AutoCertCachePath: ""
            ServerCertificate: ""
            ServerTLSKey: ""
```

The leaf key with the path Node.ClientAPI.Host translates to the environment variable BACALHAU_NODE_CLIENTAPI_HOST.

The leaf key with the path Node.ClientAPI.Port translates to the environment variable BACALHAU_NODE_CLIENTAPI_PORT.

The leaf key with the path Node.ClientAPI.TLS.AutoCert translates to the environment variable BACALHAU_NODE_CLIENTAPI_TLS_AUTOCERT.

... And so on for each leaf key in the YAML.

Branch nodes, such as Node, ClientAPI, ServerAPI, and TLS, won't have direct environment variable representations since they contain nested values. Only the leaf nodes, the keys with direct values like Host, Port, and the specific keys inside TLS, will have corresponding environment variables.

YAML can be visualized as a tree structure, and I am using the term "leaf key" rather than "leaf node" because a leaf node would typically refer to the end-point value in a tree e.g., bootstrap.production.bacalhau.org or 1234. Where as leaf key refers to the key which points to an end-point value e.g., Host, Port

> Bacalhau establishes a direct relationship between the leaf keys within the `config.yaml` file and corresponding environment variables. For leaf keys, the environment variable name is constructed by capitalizing each segment of the key, and then joining them with underscores, prefixed with `BACALHAU_`. 
> 
> For example, a YAML leaf key with the path `Node.Ipfs.Connect` translates to the environment variable `BACALHAU_NODE_IPFS_CONNECT`. 

Reviewer: ```suggestion
Bacalhau establishes a direct relationship between the value-bearing keys within the `config.yaml` file and corresponding environment variables. For these keys that have no further sub-keys, the environment variable name is constructed by capitalizing each segment of the key, and then joining them with underscores, prefixed with `BACALHAU_`. 

For example, a YAML key with the path `Node.Ipfs.Connect` translates to the environment variable `BACALHAU_NODE_IPFS_CONNECT` and is repesented in a file like:

Node:
    IPFS:
        Connect: value
        
The is no corresponding environment variable for either `Node` or `Node.IPFS`.
```

the sample probably needs ` around it, but maybe just not mentioning leaf is enough?

It's important to emphasize that this transformation applies exclusively to leaf keys. Nodes with nested values within the `config.yaml` are not represented as environment variables in this manner.

### Environments

> - Bacalhau leverages the `BACALHAU_ENVIRONMENT` environment variable to determine the specific environment configuration when initializing a repository. Notably, if a `.bacalhau` repository has already been initialized, the `BACALHAU_ENVIRONMENT` setting will be disregarded.

Reviewer: ```suggestion
- Bacalhau leverages the `BACALHAU_ENVIRONMENT` environment variable to determine the specific environment configuration when initializing a repository. Notably, if a `.bacalhau` repository has already been initialized, the `BACALHAU_ENVIRONMENT` setting will be ignored.
```

  #### 2. Staging (staging network)

  - **Environment Variable:** `BACALHAU_ENVIRONMENT=staging`
  - Configurations:
    - `Node.ClientAPI.Host`: `"bootstrap.staging.bacalhau.org"`
    - `Node.Client.API.Host`: `1234`
    - *...other configurations specific to this environment...*

  #### 3. Development (development network)

  - **Environment Variable:** `BACALHAU_ENVIRONMENT=development`
  - Configurations:
    - `Node.ClientAPI.Host`: `"bootstrap.development.bacalhau.org"`
    - `Node.Client.API.Host`: `1234`
    - *...other configurations specific to this environment...*

  #### 4. Local (private or local networks)

  - **Environment Variable:** `BACALHAU_ENVIRONMENT=local`
  - Configurations:
    - `Node.ClientAPI.Host`: `"0.0.0.0"`
    - `Node.Client.API.Host`: `1234`
    - *...other configurations specific to this environment...*

  ------

  **Note**: The above configurations provided for each environment are not exhaustive. Consult the specific environment documentation for a [comprehensive list of configurations](https://github.com/bacalhau-project/bacalhau/tree/main/pkg/config/configenv).

## Usage Examples

> **How to initialize a Bacalhau Server for a local private network**

Reviewer 2: Headings plz

Author: d832311b2c49b56db83c070394c7deaa968c6f8e
